deploy/unifi.sh

```diff
@@ -24,77 +55,155 @@ unifi_deploy() {
-      _err "unifi keystore is not found, please define DEPLOY_UNIFI_KEYSTORE"
+    _debug "Generate import pkcs12"
+    _import_pkcs12="$(_mktemp)"
+    _toPkcs "$_import_pkcs12" "$_ckey" "$_ccert" "$_cca" "$_unifi_keypass" unifi root
+    # shellcheck disable=SC2181
+    if [ "$?" != "0" ]; then
+      _err "Oops, error creating import pkcs12, please report bug to us."
       return 1
+    fi
+
+    _debug "Import into keystore: $_unifi_keystore"
+    if keytool -importkeystore \
+      -deststorepass "$_unifi_keypass" -destkeypass "$_unifi_keypass" -destkeystore "$_unifi_keystore" \
+      -srckeystore "$_import_pkcs12" -srcstoretype PKCS12 -srcstorepass "$_unifi_keypass" \
+      -alias unifi -noprompt; then
+      _debug "Import keystore success!"
+      rm "$_import_pkcs12"
     else
-      _err "It seems that the specified unifi keystore is not valid, please check."
+      _err "Error importing into Unifi Java keystore."
+      _err "Please re-run with --debug and report a bug."
+      rm "$_import_pkcs12"
       return 1
     fi
+
+    DEFAULT_UNIFI_RELOAD="${DEFAULT_UNIFI_RELOAD} ${DEFAULT_UNIFI_RELOAD:+&&} service unifi restart"
+    _info "Install Unifi Controller certificate success!"
+  elif [ "$DEPLOY_UNIFI_KEYSTORE" ]; then
+    _err "The specified DEPLOY_UNIFI_KEYSTORE='$DEPLOY_UNIFI_KEYSTORE' is not valid, please check."
+    return 1
   fi
-  if [ ! -w "$_unifi_keystore" ]; then
-    _err "The file $_unifi_keystore is not writable, please change the permission."
+
+  # Cloud Key environment (non-UnifiOS -- nginx serves admin pages) --
+  # auto-detect by file /etc/ssl/private/cloudkey.key:
+  DEFAULT_DEPLOY_UNIFI_CLOUDKEY_CERTDIR="/etc/ssl/private"
+  _cloudkey_certdir="${DEPLOY_UNIFI_CLOUDKEY_CERTDIR:-$DEFAULT_DEPLOY_UNIFI_CLOUDKEY_CERTDIR}"
+  if [ -f "${_cloudkey_certdir}/cloudkey.key" ]; then
+    _info "Installing certificate for Cloud Key Gen1 (nginx admin pages)"
+    _debug _cloudkey_certdir "$_cloudkey_certdir"
+    if [ ! -w "$_cloudkey_certdir" ]; then
+      _err "The directory $_cloudkey_certdir is not writable; please check permissions."
+      return 1
+    fi
+    # Cloud Key expects to load the keystore from /etc/ssl/private/unifi.keystore.jks.
+    # Normally /usr/lib/unifi/data/keystore is a symlink there (so the keystore was
+    # updated above), but if not, we don't know how to handle this installation:
+    if ! cmp -s "$_unifi_keystore" "${_cloudkey_certdir}/unifi.keystore.jks"; then
+      _err "Unsupported Cloud Key configuration: keystore not found at '${_cloudkey_certdir}/unifi.keystore.jks'"
+      return 1
+    fi
+
+    cp "$_cfullchain" "${_cloudkey_certdir}/cloudkey.crt"
+    cp "$_ckey" "${_cloudkey_certdir}/cloudkey.key"
+    (cd "$_cloudkey_certdir" && tar -cf cert.tar cloudkey.crt cloudkey.key unifi.keystore.jks)
+    _info "Install Cloud Key Gen1 certificate success!"
+
+    DEFAULT_UNIFI_RELOAD="${DEFAULT_UNIFI_RELOAD} ${DEFAULT_UNIFI_RELOAD:+&&} service nginx restart"
```

Comment:
Ahh... this is where it could fall apart though. An upgraded (v1 -> v2) CKG2/2+ will still have these files present (`cloudkey.{key,crt}`), but no longer runs `nginx`. This attempted restart will cause a failure.

I'm not sure what the right solution is... I'll put my thinking cap on.

Reply:
perhaps `systemctl -q is-active nginx && systemctl restart nginx`

Reply:
`systemctl -q is-active nginx` looks good on my CKG1.

Maybe a good idea to guard each of the three restart commands with this. (Would an upgraded CKG2 have a leftover /usr/lib/unifi/data/keystore that would trigger an unwanted `service unifi restart`?)

Reply:
Also, on CKG2+, which runs `unifi-protect`, it might make sense to restart that as well, when `cloudkey.crt` is replaced. I cannot verify anymore, but I suspect something like this addition would work:

```sh
if [ -L /srv/unifi-protect/data/unifi-protect.crt ];then
  if [ "$(realpath /srv/unifi-protect/data/unifi-protect.crt)" = "${_cloudkey_certdir}/cloudkey.crt" ]; then
    DEFAULT_UNIFI_RELOAD="${DEFAULT_UNIFI_RELOAD} ${DEFAULT_UNIFI_RELOAD:+&&} systemctl -q is-active unifi-protect && systemctl restart unifi-protect"
  fi
fi
```

I have asked a friend with a CKG2+ still on v1 to check if that file is a symlink or not - it no longer is on v2.

Reply:
> `systemctl -q is-active nginx` looks good on my CKG1.

Sweet, that's handy

> Maybe a good idea to guard each of the three restart commands with this.

Yes, it would make things safer for sure.

> (Would an upgraded CKG2 have a leftover /usr/lib/unifi/data/keystore that would trigger an unwanted `service unifi restart`?)

Very, very good question. This is something I've actually posted about recently on the UBNT community forums. Technically, it's still used - not by an end user using the controller admin interface, but it is used as the cert for a Guest network captive portal... I suspect it's still worthwhile to leave as you have it - being updated and `unifi` restarted.

Reply:
OK, for each of the services unifi, nginx, and unifi-core, I believe the script will now try to:
* If the cert files for that service are present (whether or not that service is still in use), update with the new cert
* If the service is active, restart it to pick up the updated certs

Reply:
Your patch looks great, nice work.

Reply:
Thanks for all the help.

Reply:
No problem - would love to see a restart of `unifi-protect` added if I can confirm the cert link state for you.. just an extra restart to add when replacing `/etc/ssl/private/cloudkey.crt` when conditions are met. I'll keep you updated.